crates/extensions/tedge_mqtt_ext/src/test_helpers.rs

```diff
@@ -48,14 +33,88 @@ pub async fn assert_received_includes_json<I, S>(
 {
     for expected_msg in expected.into_iter() {
         let message = messages.recv().await.expect("MQTT channel closed");
-        assert_eq!(message.topic, Topic::new_unchecked(expected_msg.0.as_ref()));
-        let payload = serde_json::from_str::<serde_json::Value>(
-            message.payload_str().expect("non UTF-8 payload"),
-        )
-        .expect("non JSON payload");
-        assert_json_include!(
-            actual: payload,
-            expected: expected_msg.1
-        );
+        assert_message_includes_json(&message, expected_msg);
+    }
+}
+
+pub fn assert_messages_contains_str<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
+where
+    I: IntoIterator<Item = (S, S)>,
+    S: AsRef<str> + Debug,
+{
+    let mut expected = expected.into_iter();
+    loop {
+        match (messages.is_empty(), expected.next()) {
+            (false, Some(expected_msg)) => {
+                let message = messages.remove(0);
+                assert_message_contains_str(&message, expected_msg);
+            }
+            (true, Some(expected_msg)) => {
+                panic!(
+                    "Input messages exhausted while expecting: {:?}",
+                    expected_msg
+                )
+            }
+            _ => break,
+        }
     }
 }
+
+pub fn assert_messages_includes_json<I, S>(messages: &mut Vec<MqttMessage>, expected: I)
+where
+    I: IntoIterator<Item = (S, serde_json::Value)>,
+    S: AsRef<str>,
+{
+    let mut expected = expected.into_iter();
+    loop {
+        match (messages.is_empty(), expected.next()) {
+            (false, Some(expected_msg)) => {
+                let message = messages.remove(0);
+                assert_message_includes_json(&message, expected_msg);
+            }
+            (true, Some(expected_msg)) => {
+                panic!(
+                    "Input messages exhausted while expecting message on topic: {:?} with payload: {:?}",
+                    expected_msg.0.as_ref(), expected_msg.1
+                )
+            }
+            _ => break,
+        }
+    }
+}
+
+pub fn assert_message_contains_str<S>(message: &Message, expected: (S, S))
+where
+    S: AsRef<str> + Debug,
```

Comment:
It seems a bit arbitrary to me here that we accept anything that implements `AsRef<str>` here for the topic and the payload, but the two have to be the same concrete type.

```suggestion
pub fn assert_message_contains_str(message: &Message, expected: (impl AsRef<str>, impl AsRef<str>))
```

would resolve this (as would adding a second type parameter. Given this is test code, I would argue we maybe don't need the complexity of `impl AsRef<str>` at all, most assertions are going to be static (so `&str` is what we're passing in anyway), and if we have `String`, it's trivial to convert to `&str` manually.

Reply:
Yeah, they didn't have to be the same type as implied by `(S, S)`. I did that just to avoid the overkill of another generic parameter, as the inputs were always static `str` values. So, yes, `&str` probably is the best/simplest choice here.